src/devices/src/virtio/net/device.rs

```diff
@@ -257,84 +276,94 @@ impl Net {
+                break;
+            }
 
-                    match write_result {
-                        Ok(()) => {
-                            METRICS.net.rx_count.inc();
-                            write_count += source_slice.len();
-                        }
-                        Err(e) => {
-                            error!("Failed to write slice: {:?}", e);
-                            if let GuestMemoryError::PartialBuffer { completed, .. } = e {
-                                write_count += completed;
-                                METRICS.net.rx_partial_writes.inc();
-                            } else {
-                                METRICS.net.rx_fails.inc();
-                            }
-                            break;
-                        }
-                    };
+            if !descriptor.is_write_only() {
+                result = Err(FrontendError::ReadOnlyDescriptor);
+                break;
+            }
 
-                    if write_count >= self.rx_bytes_read {
-                        break;
-                    }
-                    next_desc = desc.next_descriptor();
+            let len = std::cmp::min(frame_slice.len(), descriptor.len as usize);
+            match mem.write_slice(&frame_slice[..len], descriptor.addr) {
+                Ok(()) => {
+                    METRICS.net.rx_count.inc();
+                    frame_slice = &frame_slice[len..];
                 }
-                None => {
-                    warn!("Receiving buffer is too small to hold frame of current size");
-                    METRICS.net.rx_fails.inc();
+                Err(e) => {
+                    error!("Failed to write slice: {:?}", e);
+                    match e {
+                        GuestMemoryError::PartialBuffer { .. } => &METRICS.net.rx_partial_writes,
+                        _ => &METRICS.net.rx_fails,
+                    }
+                    .inc();
+                    result = Err(FrontendError::GuestMemory(e));
                     break;
                 }
-            }
-        }
+            };
 
-        rx_queue
-            .add_used(mem, head_index, write_count as u32)
-            .unwrap_or_else(|e| {
-                error!("Failed to add available descriptor {}: {}", head_index, e);
-            });
+            maybe_next_descriptor = descriptor.next_descriptor();
+        }
+        if result.is_ok() && !frame_slice.is_empty() {
+            warn!("Receiving buffer is too small to hold frame of current size");
+            METRICS.net.rx_fails.inc();
+            result = Err(FrontendError::DescriptorChainTooSmall);
+        }
 
-        // Mark that we have at least one pending packet and we need to interrupt the guest.
+        // Mark the descriptor chain as used. If an error occurred, skip the descriptor chain.
+        let used_len = if result.is_err() { 0 } else { frame_len as u32 };
+        queue.add_used(mem, head_index, used_len).map_err(|e| {
```

Comment:
If we provide a descriptor chain with 0 used bytes, the driver will simply skip it. This is not a waste of guest cycles. We actually have to notify the guest that the descriptor chain has been used, because otherwise the guest might not provide new descriptor chains for us. It will think that we already have enough available.

This is covered in the unit tests. But there we don't get to see the actual driver behavior. I can't think of a way to trigger it in the integration tests.